File: leetcode_problems/typescript/src/lc526_beautiful_arrangement.ts
function countArrangement(n: number): number {
  function backtrack(cur: number) {
    if (cur > n) {
      count += 1;
      return;
    }

    for (let i = 0; i < n; i++) {
      if (arr[i] !== 0) {
        continue;
      }
      if (cur % (i + 1) !== 0 && (i + 1) % cur !== 0) {
        continue;
      }

      arr[i] = cur;
      backtrack(cur + 1);
      arr[i] = 0;
    }
  }

  let count = 0;
  let arr: number[] = new Array(n).fill(0);
  backtrack(1);
  return count;
}

File: leetcode_problems/cpp/lc526_beautiful_arrangement.cpp
class Solution
{
public:
    void backtrack(int cur, int *arr, int n, int &count)
    {
        if (cur > n)
        {
            ++count;
        }

        for (int i = 0; i < n; i++)
        {
            if (arr[i] != 0)
            {
                continue;
            }
            if (cur % (i + 1) != 0 && (i + 1) % cur != 0)
            {
                continue;
            }

            arr[i] = cur;
            backtrack(cur + 1, arr, n, count);
            arr[i] = 0;
        }
    }

    int countArrangement(int n)
    {
        int count = 0;
        int *arr = new int[n]();
        backtrack(1, arr, n, count);
        delete[] arr;
        return count;
    }
};
